### PR TITLE
refactor(connector): [Authorizedotnet] Enhance currency Mapping with ConnectorCurrencyCommon Trait

### LINKED ISSUE #2218: [FEATURE]: [Authorizedotnet] Currency Unit Conversion
### :memo: Feature Description

- Each currency can be described in terms of base or minor units. 
- For instance, when dealing with USD currency, the Cent is considered as the minor unit, while the Dollar is considered the base unit. 
- In Hyperswitch, the amount value is expected to be always provided in minor units. 
- For example: In USD, If the amount is 1 in base unit (Dollar), then it will be equivalent to 100 in minor units (Cent)
- Some of the connectors integrated require the amount to be converted to a desirable unit before being passed to them. 
- We have functions `to_currency_base_unit` and `to_currency_lower_unit` in place to convert the minor unit amount to its decimal equivalent and vice versa. These conversions are handled explicitly based on the connector.

### :hammer: Possible Implementation

- ConnectorCommon trait have been implemented for the connector. 
- This trait contains `get_currency_unit` method. This method needs to be implemented. 
- It will define what type of conversion needs to be done during `connector_router_data` creation. Concurrently handle the creation of `connector_router_data` to be passed for the connector request body creation.
- You can check this PR for further reference https://github.com/juspay/hyperswitch/pull/2196

:bookmark: Note: All the changes needed should be contained within `hyperswitch/crates/router/src/connector/`

### :package: Have you spent some time checking if this feature request has been raised before?

- [X] I checked and didn't find a similar issue

### :package: Have you read the Contributing Guidelines?

- [X] I have read the [Contributing Guidelines](https://github.com/juspay/hyperswitch/blob/main/docs/CONTRIBUTING.md)

### PR DESCRIPTION
**Pull request by Azanul**
## Type of Change

- [ ] Bugfix
- [ ] New feature
- [ ] Enhancement
- [X] Refactoring
- [ ] Dependency updates
- [ ] Documentation
- [ ] CI/CD

### Additional Changes

- [ ] This PR modifies the API contract
- [ ] This PR modifies the database schema
- [ ] This PR modifies application configuration/environment variables

## Motivation and Context

Fixes #2218 

## Checklist

- [X] I formatted the code `cargo +nightly fmt --all`
- [ ] I addressed lints thrown by `cargo clippy`
- [X] I reviewed the submitted code
- [ ] I added unit tests for my changes where possible
- [ ] I added a [CHANGELOG](/CHANGELOG.md) entry if applicable

### CODE REVIEW

File: crates/router/src/connector/authorizedotnet.rs [hunk 1]
```diff
@@ -142,7 +146,14 @@ impl ConnectorIntegration<api::Capture, types::PaymentsCaptureData, types::Payme
         &self,
         req: &types::PaymentsCaptureRouterData,
     ) -> CustomResult<Option<types::RequestBody>, errors::ConnectorError> {
-        let connector_req = authorizedotnet::CancelOrCaptureTransactionRequest::try_from(req)?;
+        let connector_router_data = authorizedotnet::AuthorizedotnetRouterData::try_from((
+            &self.get_currency_unit(),
+            req.request.currency,
+            req.request.payment_amount,
```
Comment (srujanchikke) [inline crates/router/src/connector/authorizedotnet.rs:152]:
```suggestion
            req.request.amount_to_capture,
```
This is capture flow , amount should be capture amount

File: crates/router/src/connector/authorizedotnet.rs [hunk 2]
```diff
@@ -407,7 +425,14 @@ impl ConnectorIntegration<api::Void, types::PaymentsCancelData, types::PaymentsR
         &self,
         req: &types::PaymentsCancelRouterData,
     ) -> CustomResult<Option<types::RequestBody>, errors::ConnectorError> {
-        let connector_req = authorizedotnet::CancelOrCaptureTransactionRequest::try_from(req)?;
+        let connector_router_data = authorizedotnet::AuthorizedotnetRouterData::try_from((
+            &self.get_currency_unit(),
+            req.request.currency.unwrap(),
```
Comment (srujanchikke) [inline crates/router/src/connector/authorizedotnet.rs:430]: amount is not required for Authorizedotnet cancel flow  . You can revert changes for void flow . And also unwrap shouldn't be used as application goes to panic state instead throw appropriate error.

File: crates/router/src/connector/authorizedotnet/transformers.rs [hunk 3]
```diff
@@ -661,21 +729,22 @@ impl<F> TryFrom<&types::RefundsRouterData<F>> for CreateRefundRequest {
...
-        let merchant_authentication = AuthorizedotnetAuthType::try_from(&item.connector_auth_type)?;
+        let merchant_authentication =
+            AuthorizedotnetAuthType::try_from(&item.router_data.connector_auth_type)?;
 
         let transaction_request = RefundTransactionRequest {
             transaction_type: TransactionType::Refund,
             amount: utils::to_currency_base_unit_asf64(
-                item.request.refund_amount,
-                item.request.currency,
+                item.router_data.request.refund_amount,
```
Comment (srujanchikke) [inline crates/router/src/connector/authorizedotnet/transformers.rs:738]:
```suggestion
                amount : item.amount.clone(),
```

File: crates/router/src/connector/authorizedotnet/transformers.rs [hunk 4]
```diff
@@ -263,26 +299,37 @@ impl From<enums::CaptureMethod> for AuthorizationType {
...
+                .request
+                .capture_method
+                .map(|c| AuthorizationIndicator {
+                    authorization_indicator: c.into(),
+                });
         let transaction_request = TransactionRequest {
-            transaction_type: TransactionType::from(item.request.capture_method),
-            amount: utils::to_currency_base_unit_asf64(item.request.amount, item.request.currency)?,
+            transaction_type: TransactionType::from(item.router_data.request.capture_method),
+            amount: utils::to_currency_base_unit_asf64(
```
Comment (srujanchikke) [inline crates/router/src/connector/authorizedotnet/transformers.rs:320]:
```suggestion
            amount: item.amount.clone(),
```

File: crates/router/src/connector/authorizedotnet/transformers.rs [hunk 5]
```diff
@@ -313,19 +369,28 @@ impl TryFrom<&types::PaymentsCancelRouterData> for CancelOrCaptureTransactionReq
...
     type Error = error_stack::Report<errors::ConnectorError>;
-    fn try_from(item: &types::PaymentsCaptureRouterData) -> Result<Self, Self::Error> {
+    fn try_from(
+        item: &AuthorizedotnetRouterData<&types::PaymentsCaptureRouterData>,
+    ) -> Result<Self, Self::Error> {
         let transaction_request = TransactionVoidOrCaptureRequest {
             amount: Some(utils::to_currency_base_unit_asf64(
-                item.request.amount_to_capture,
-                item.request.currency,
+                item.router_data.request.amount_to_capture,
```
Comment (srujanchikke) [inline crates/router/src/connector/authorizedotnet/transformers.rs:381]: amount can be fetched directly from AuthorizedotnetRouterData as conversion is already happening in `AuthorizedotnetRouterData` tryfrom
```suggestion
                item.amount.clone(),
```

File: crates/router/src/connector/authorizedotnet/transformers.rs (hunk 4 above)
Comment (Azanul) [reply crates/router/src/connector/authorizedotnet/transformers.rs:320]: amount in TransactionRequest is a f64 and amount in AuthorizedotnetRouterData is a String

File: crates/router/src/connector/authorizedotnet/transformers.rs (hunk 5 above)
Comment (Azanul) [reply crates/router/src/connector/authorizedotnet/transformers.rs:381]: amount in TransactionVoidOrCaptureRequest is a f64 and amount in AuthorizedotnetRouterData is a String

File: crates/router/src/connector/authorizedotnet/transformers.rs (hunk 3 above)
Comment (Azanul) [reply crates/router/src/connector/authorizedotnet/transformers.rs:738]: amount in RefundTransactionRequest is a f64 and amount in AuthorizedotnetRouterData is a String

File: crates/router/src/connector/authorizedotnet.rs (hunk 2 above)
Comment (Azanul) [reply crates/router/src/connector/authorizedotnet.rs:430]: CancelOrCaptureTransactionRequest is used for cancel request and capture request both, if I revert for cancel then I'll have to do the same for capture request

**Comment by Azanul** [comment]
@srujanchikke Please ignore my review comments, I looked into it and have respectively fixed the issues.

**Review by srujanchikke — APPROVED** [review]
Looks good to me!

**Review by SanchithHegde — APPROVED** [review]
Looks good to me!

Thanks for the PR, @Azanul!

**Comment by VedantKhairnar** [comment]
Hey @Azanul  👋,
Thanks a bunch for all your contributions! We've got a little something for you to show our appreciation.
Just take a moment to fill out this [form](https://forms.gle/yqbtjFuP6oxLaBKN8), and get ready for some awesome swag coming your way.
Thanks!

**Pull request merged**